### PR TITLE
ENH/WIP: link electron energy, calculate photon energy

### PR DESCRIPTION
**Pull request by klauer**
Pending:
* Electron energy PV names for both undulator lines

### CODE REVIEW

File: Arbiter/ArbiterPLC/POUs/Accelerator/FB_Hgvpu.TcPOU [hunk 1]
```diff
@@ -91,6 +91,7 @@ END_VAR
 VAR CONSTANT
     iLowBound  : UDINT := 24;
     iHighBound : UDINT := 46;
+    fPeriod_mm : LREAL := 26.0;
```
Comment (slacAWallace) [inline Arbiter/ArbiterPLC/POUs/Accelerator/FB_Hgvpu.TcPOU:94]: I've been exporting constants to EPICS for the PMPS, it helps. I recommend it here.

File: Arbiter/ArbiterPLC/POUs/Accelerator/FB_Hgvpu.TcPOU [hunk 2]
```diff
@@ -120,13 +121,21 @@ END_VAR
     fbSegment[45] := ADR(fbSegment_45);
     fbSegment[46] := ADR(fbSegment_46);
 
+
+    FOR iIndex := iLowBound TO iHighBound DO
+            IF fbSegment[iIndex] <> 0 THEN
```
Comment (slacAWallace) [inline Arbiter/ArbiterPLC/POUs/Accelerator/FB_Hgvpu.TcPOU:126]: Legit.

File: Arbiter/ArbiterPLC/POUs/FB_StopperWatcher.TcPOU [hunk 3]
```diff
@@ -46,7 +46,7 @@ q_StopperOUT_Relay := i_StopperOutLS;
 q_StopperIN_Relay := i_StopperInLS;
 
 // Update current beam parameters
-stCurrentBP.aStoppers[]]></ST>
+//stCurrentBP.aStoppers[]]></ST>
```
Comment (slacAWallace) [inline Arbiter/ArbiterPLC/POUs/FB_StopperWatcher.TcPOU:49]: O_o

File: Arbiter/ArbiterPLC/POUs/Accelerator/F_CalculatePhotonEnergy.TcPOU [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    fUndulatorPeriod_mm : LREAL;
+    (* Unitless undulator K parameter / strength *)
+    fUndulatorStrength : LREAL;
+END_VAR
+
+VAR
+    fDenominator: LREAL;
+END_VAR
+
+VAR CONSTANT
```
Comment (slacAWallace) [inline Arbiter/ArbiterPLC/POUs/Accelerator/F_CalculatePhotonEnergy.TcPOU:18]: Same as above. We'll put them in the diagnostic display for photon energy calc. Warm and fuzzy feelings all around.

File: Arbiter/ArbiterPLC/POUs/Accelerator/F_CalculatePhotonEnergy.TcPOU [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    '''
+    m_e = 0.0005109989461  # GeV [electron rest mass/energy]
+    h = 6.62607004e-34  # Js [Planck's constant]
+    e = 1.6021766208e-19  # C [electron charge]
+    c = 299792458  # m/s, speed of light
+    return ((2. * (electron_energy / m_e) ** 2 * h * c) /
+            (e * period * 1e-3 * (1 + k ** 2 / 2.)))
+
+*)
+fDenominator := (e * fUndulatorPeriod_mm * 1E-3 * (1.0 + EXPT(fUndulatorStrength, 2.0) / 2.0));
```
Comment (slacAWallace) [inline Arbiter/ArbiterPLC/POUs/Accelerator/F_CalculatePhotonEnergy.TcPOU:58]: A div-by-zero here is super unlikely, but it is good practice, especially with these systems to guard against any possibility like this. Zach's idea of using a max function against a tiny float was a good way of doing it.

**Comment by klauer** [comment]
Thanks, @slacAWallace. Will fix as requested once I get the PVs.

**Comment by slacAWallace** [comment]
@klauer We should aim for next Tuesday to integrate this PR and deploy it. It's our last PAMM before HXR beam commissioning.

**Comment by klauer** [comment]
Accelerator says the PVs don't exist and are working to help us make them. Whether that will meet the Tuesday deadline or not is not up to me...

**Comment by slacAWallace** [comment]
Understood. Can we at least agree on a PV name to start so our side will be ready without another deployment?

**Comment by klauer** [comment]
Jeremy agreed to help decide a PV name. Will make sure we get it by Monday to implement/deploy.

In the meantime, who can we have verify that we're not doing something dumb with the calculations? Python code this calculation is based on shows the following, which may or may not be sensible given my lack of physics knowledge:

image

**Pull request merged**